docs/host-file-isolation.md

```diff
@@ -0,0 +1,142 @@
+management of a VM.  We expect that every VM will use the same version of
+Firecracker and Jailer.  While we do not expect there to be a mechanism by which
+a running VM can write to either of these files, we do believe it's important to
+enforce that expectation external to Firecracker and Jailer.
+
+## Snapshotter and content store
+
+The snapshotter and content store are technically backed by files, but
+incorporate enough difference to warrant a separate document.
+
+## Techniques
+
+Any of the file components that need to be shared can be done so through several
+basic ways:
+
+* Reusing the same backing storage
+* Copying (duplicating the content)
+* Copy-on-write
+
+There are a variety of techniques to accomplish each of these.
+
+### Reusing storage
+
+* Opening the same file
+* Creating a hard-link to the same file and opening it
+* Creating a symbolic-link to the same file and opening it
+* Bind-mounting the file elsewhere on the filesystem and opening it
+* Opening an existing file descriptor from the `/proc` filesystem
+
+All of these allow for mutation of the original storage, unless protected by
+another mechanism layered on top.  These are all very efficient mechanisms; they
+do not incur additional latency or storage space.  However, we still need a
+mechanism for preventing mutation.
+
+### Copying
+
+Copying ahead of launching a microVM provides assurance that the original file
+will not be modified, as the original backing storage is not opened and not
+accessible.  However, it has the downsides of being expensive in terms of time
+(a latency impact on pulling an image or launching a new microVM) and expensive
+in terms of space (scaling linearly with the number of microVMs we launch).
+
+Copies can be performed to local, durable storage, to network-attached storage,
+or to memory with a tmpfs or memfd.
+
+### Copy-on-write
+
+Copy-on-write techniques provide mechanisms for attempting to achieve the same
+efficiency as reusing storage but allowing the safety of preventing modification
+of the original storage.  There are a variety of copy-on-write solutions:
+
+* Overlay filesystem (file-based copy-on-write storage, copy-up is performed
+  when a file is opened for write)
+* Devicemapper thin devices (block-based copy-on-write, copy-up is performed
+  when a block is written)
+* Filesystem-integrated copy-on-write (ZFS, BTRFS, XFS, etc)
+
+Overlay has broad support as it is integrated with the Linux kernel and
+Firecracker already requires a new-enough kernel.  Overlay is also simple to set
+up.  We are already using devicemapper as our snapshotter, even though it is
+more challenging to set up.  Filesystem-integrated copy-on-write approaches
+require a filesystem that supports the feature; we do not currently require a
+specific filesystem to run firecracker-containerd.
+
+### Preventing mutation
+
+Linux provides different capabilities for preventing mutation to files:
```

Comment:
You could add "mounting filesystem as read-only" to this list I think

Reply:
Done